### ISSUE TITLE
Feature request: combine FERAL with other frame-wise features (pose, sensors, …)

### ISSUE THREAD

**Issue by Akseli-Ilmanen** [Skovorp/feral#21]
Hi, 

Thanks for creating FERAL, really cool tool! 

I participated in the recent `movement` community call, and found the discussion about how FERAL+SAM can be used for multi-animal scenarios super interesting. 

### Context

The paper says FERAL "can operate either alongside or independently of existing pose-based
frameworks". As far as I can tell from the code, that combination happens *outside* FERAL:
`FeralModel.forward` takes only frames, and there is no way to give the model, or a downstream
model, anything else. It would be great to have a supported way to do this.

Our use case, for concreteness:

- Single-animal tool-use sequences in carrion crows (previous work: [Moll et al., 2025](https://www.sciencedirect.com/science/article/pii/S0960982225011005)), filmed at **200 fps**, with 3D pose (beak and tool tip), an accelerometer, and per-frame features derived from them (velocities, distances to landmarks, speed changepoints).
- Densely labelled state behaviours: ~7,700 curated segments, 28 classes, and **short**
  (median 0.18 s, 5th–95th percentile 0.10–0.83 s), about 13 behaviours per ~7 s trial, in a
  fairly stereotyped order.
- We care about boundaries at frame precision, so we score segmental F1 at IoU 0.5/0.75/0.9, not
  only frame-level mAP/F1.

### Why frame-wise features: what we saw with a frozen video backbone

Before trying FERAL, we trained action segmentation models on video features from a **frozen**
network: S3D features extracted with [video_features](https://github.com/v-iashin/video_features).
Adding a small set of per-frame **changepoint features** to those S3D features (moments where the
movement changes, expanded into a few smooth per-frame signals) improved segmentation a lot,
e.g. F1@90 from roughly **40 % to 60 %**. The video features carry *what* is happening; the
changepoint features tell the model *exactly when* a behaviour starts and ends.

In our data these changepoint features are derived from pose (see image below). The scenario I find most
interesting for FERAL, though, is combining it with **per-animal IMU sensors**: changepoints
computed from each animal's accelerometer, with a workflow like
[this one](https://akseli-ilmanen.github.io/ethograph/advanced/changepoints/overview.html),
fed to FERAL alongside the video. That needs no pose estimation at all, which fits FERAL's
"direct video-to-behaviour" idea, and it gives each individual its own timing signal even in
group recordings. It would be great to test on datasets that record video and body-worn sensors
together, such as BirdPark (https://pubmed.ncbi.nlm.nih.gov/41244214/).

<img width="504" height="271" alt="Image" src="https://github.com/user-attachments/assets/275cf7ca-66b7-4fe6-afd2-e85f9f71c6e1" />

On our split, FERAL `lite` (video only) recognises the behaviours reasonably well but places
their boundaries poorly compared with a temporal convolutional model on pose + changepoint
features:

| Test set | F1@50 | F1@75 | F1@90 |
|---|---|---|---|
| FERAL lite, video only (`chunk_step` 6, 80 % inference overlap) | 76.8 | 38.1 | 10.5 |
| C2F-TCN on pose + accelerometer + changepoint features | 88.2 | 77.3 | 62.1 |

Note my goal is not for FERAL to beat the C2F model on our hand-crafted features. That is probably a ceiling. But using our crow dataset, I would like to test if FERAL with changepoint features significantly improves the F1@90 10.5 score. If it does, this would make the FERAL + IMU sensor changepoints idea more tangible. Besides changepoints, I am currently testing how a lower `chunk_step` will help with F1@90. With 200 fps, this comes at the expense of a very small chunk window.

### Option A — pass frame-wise features into the attention head

**Note**: These two suggestions are drafted with Claude. Please correct anything that misrepresents how FERAL works.

FERAL's head already has one learnable query per frame of the chunk (`predict_per_item` = 64),
so the time axis of the pooled output lines up with the input frames. A user-supplied `(T, F)`
array per video (same frame count as the video, like the labels) could be gathered at the chunk's
`frame_ids` and injected at either of two points:

1. **Into the queries, before cross-attention:** `query_t = learned_query_t + MLP(features_t)`.
   The features can then steer *where* in the video each frame's query looks.
2. **After pooling, before BatchNorm + Linear:** `logits_t = Linear(BN([pooled_t ; MLP(features_t)]))`.
   This is the simplest to implement and is essentially late fusion inside the model.

This could be a small, opt-in change, e.g. `data.feature_dir` (one `.npy` per video) plus
`model.feature_fusion: query | concat`, with the dataset returning the gathered features next to
the frames. Normalisation (z-scoring from the train split) could live in the dataset.

Video and features are trained together in one model, so nothing leaks between stages (see
Option B). One limitation: at `chunk_step > 1` the features are only seen at the sampled frames,
so the boundary precision they can add is still bounded by the stride.

### Option B — export FERAL's per-frame embeddings for a downstream temporal model

Alternatively, a way to train FERAL as usual and then **export the per-frame embeddings** (the
attention-pooled `(64, D)` output, ensembled over overlapping chunks the same way the
probabilities are), e.g. `feral infer ... --save-embeddings` writing one `(T, D)` array per video.

These could then be concatenated with other features and used to train a dedicated action
segmentation model (MS-TCN, ASFormer, C2F-TCN, the [DLC2Action](https://github.com/amathislab/DLC2action) architectures, …). Those models learn multi-resolution temporal dependencies over whole trials, which helps precisely where a
fixed 64-frame chunk struggles: at high frame rates, 64 frames cover only 0.32–2 s depending on
the stride, while behaviour sequences and their ordering span many seconds. It needs no change to
FERAL's model or training, only an output option.

**The catch is label leakage, which is a real argument for Option A.** A fine-tuned FERAL has
seen the labels of its training videos, so its embeddings are more reliable on those videos than
on unseen ones. A downstream model trained on those embeddings learns to trust them more than it
should, and underperforms on test videos where FERAL is less certain. Avoiding that requires
cross-fitting: split the training videos into K folds, train K FERAL models, and export each
fold's embeddings from the model that did not train on it. That is K times the training cost + a config nightmare.
A frozen backbone (like S3D above) has no such problem, but it also gives up FERAL's fine-tuning.

### Questions

- Does Option A fit how you see the head evolving, or is it out of scope for FERAL?
- For Option B, is the pooled embedding the representation you would recommend, or an earlier
  layer? And would you expect cross-fitting to be necessary in practice?
- Have you tried something like either internally? The "alongside pose" sentence in the paper
  suggests you have thought about it.

**Comment by Skovorp** [comment]
Hi!

That’s a really cool application, we haven’t seen anything like this. We haven’t tried anything like this ourselves, as we’ve never come across datasets that had both videos and other features.

By “alongside pose,” we meant that you can do pose estimation first and then crop/rotate the video to contain only the part that you’re interested in, for example, just the head of the animal. We’ve seen people get good results with these kinds of pipelines on bee videos.

Considering your suggestion, I think option A is out of scope for FERAL. We want to keep the tool relatively simple.

Option B sounds good, we could definitely do something like this. I think using the embeddings that we get after attention pooling would be a good option. You’d probably also average the corresponding frame embeddings between overlapping chunks.

Would you be interested in doing a pull request for this? We can Zoom and discuss the details if you’re interested.

**Comment by Akseli-Ilmanen** [comment]
Hi @Skovorp 

> Option B sounds good
> Would you be interested in doing a pull request for this? We can Zoom and discuss the details if you’re interested.

This sounds great. I just opened up a PR [here](https://github.com/Akseli-Ilmanen/feral/pull/1). This is my first PR (co-authored with Claude), so very happy to get some feedback! Feel free to drop me an email: akseli.ilmanen@gmail.com, and we can arrange a time to discuss.

## Changepoints work!

Also, with my fork, I did some benchmarking, looking very promising:

**The stride was the problem** My first FERAL run used `chunk_step 6` (2 s of context per 64-frame chunk at 200 fps), which is the default I got from scaling the paper's 25 fps setup. Re-training with `chunk_step 2` (0.64 s of context, every second frame seen) changed the picture completely: on F1@50 FERAL alone now beats the C2F-TCN on our hand-crafted features (3D pose, accelerometer, distances, changepoints), from video only. I am genuinely impressed; for a 0.18 s median behaviour I did not expect a video model to get there.

**Boundaries are still the gap, and changepoint features close it.** Using the new `--save_embeddings` export (Option B), a C2F-TCN reads FERAL's 768-d per-frame embedding as its input, with and without our changepoint features. Four sessions train, one whole session (136 trials) is held out for testing; neither FERAL nor the C2F-TCN ever saw it. Same train/val trials, seed and post-processing in every row. Post-processed scores:

| model | input | frame F1 | F1@50 | F1@75 | F1@90 |
|---|---|---|---|---|---|
| C2F-TCN | FERAL embedding | 92.2 | 91.5 | 73.0 | 29.2 |
| C2F-TCN | FERAL embedding + changepoint features | 93.7 | 91.5 | 77.6 | 58.1 |

F1@50 does not move when the changepoint columns are added: the embedding already knows *what* is happening. F1@90 doubles once the model is also told *where* the movement speed dips. That is the one thing the embedding cannot carry on its own: at `chunk_step 2` it never sees a boundary finer than the stride, and the attention pooling averages over the chunk, so the onset lands within a few frames but not on the frame.

My hunch: FERAL with IMU sensor changepoint features == the new way to do action segmentation in naturalistic set-ups! :)

**Edit**: Also when FERAL to export embeddings and then train C2F-TCN on those, one could even try `chunk_step 1`, despite having 200 fps. The embeddings would only have 0.32s of context but C2F-TCN could then bridge those across many seconds. (Untested)

**Edit 2**: Btw, in case you're not familiar with segmental F1 scores. Here is a little diagram that helps showing what would be counted as a TP given different IoU criterions:

<img width="379" height="304" alt="Image" src="https://github.com/user-attachments/assets/51e69039-15ed-4144-ba91-739d9f4af110" />

**Comment by Akseli-Ilmanen** [comment]
Hi @Skovorp,

> This sounds great. I just opened up a PR [here](https://github.com/Akseli-Ilmanen/feral/pull/1.)

Ooopsie, I set up the PR on my fork not on `https://github.com/Skovorp/feral`. I fixed that now by opening PR #22. 

Best
Akseli
